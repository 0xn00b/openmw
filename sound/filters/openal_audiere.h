#ifndef MANGLE_FFMPEG_OPENAL_H
#define MANGLE_FFMPEG_OPENAL_H

#include "input_filter.h"
#include "../sources/audiere_source.h"
#include "../outputs/openal_out.h"

namespace Mangle {
namespace Sound {

/// A InputFilter that adds audiere decoding to OpenAL. Audiere has
/// it's own output, but OpenAL sports 3D and other advanced features.
class OpenAL_Audiere_Factory : public InputFilter
{
 public:
  OpenAL_Audiere_Factory()
    {
      set(new OpenAL_Factory,
          new AudiereLoader);
    }
  ~OpenAL_Audiere_Factory()
    {
      delete snd;
      delete inp;
    }
};

}}
#endif
