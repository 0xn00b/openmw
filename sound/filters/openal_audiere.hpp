#ifndef MANGLE_FFMPEG_OPENAL_H
#define MANGLE_FFMPEG_OPENAL_H

#include "input_filter.hpp"
#include "../sources/audiere_source.hpp"
#include "../outputs/openal_out.hpp"

namespace Mangle {
namespace Sound {

/// A InputFilter that adds audiere decoding to OpenAL. Audiere has
/// it's own output, but OpenAL sports 3D and other advanced features.
class OpenAL_Audiere_Factory : public InputFilter
{
 public:
  OpenAL_Audiere_Factory()
    {
      set(SoundFactoryPtr(new OpenAL_Factory),
          SampleSourceLoaderPtr(new AudiereLoader));
    }
};

}}
#endif
