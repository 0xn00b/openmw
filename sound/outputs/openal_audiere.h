#ifndef MANGLE_FFMPEG_OPENAL_H
#define MANGLE_FFMPEG_OPENAL_H

#include "input_filter.h"
#include "input_audiere.h"
#include "output_openal.h"

namespace Mangle {
namespace Sound {

/// A InputFilter that adds audiere decoding to OpenAL. Audiere has
/// it's own output, but OpenAL sports 3D and other advanced features.
class OpenAL_Audiere_Manager : public InputFilter
{
 public:
  OpenAL_Audiere_Manager()
    {
      set(new OpenAL_Manager,
          new AudiereInput);
    }
  ~OpenAL_Audiere_Manager()
    {
      delete snd;
      delete inp;
    }
};

}}
#endif
