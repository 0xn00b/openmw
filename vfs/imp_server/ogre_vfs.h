#ifndef MANGLE_VFS_OGRECLIENT_H
#define MANGLE_VFS_OGRECLIENT_H

#include <OgreArchive.h>

namespace Mangle {
namespace VFS {

/** @brief An interface into the OGRE VFS system.

    This class does not wrap a single Ogre::Archive, but rather the
    entire resource set available to Ogre. You can use this class to
    tap into all paths, Zip files, custom archives on so on that have
    been inserted into Ogre as resource locations.

    This class is currently a work in progres, it will not compile as
    it stands.

    This has been built and tested against OGRE 1.6.2. You might have
    to make your own modifications if you're working with newer (or
    older) versions.
 */
class OgreVFS : public VFS
{
 public:
  OgreVFS()
    {
      hasFind = true;
      isCaseSensitive = true;
    }

  /// Open a new data stream. Deleting the object should be enough to
  /// close it.
  virtual Stream::InputStream *open(const std::string &name);

  /// Check for the existence of a file
  virtual bool isFile(const std::string &name) const;

  /// Check for the existence of a directory
  virtual bool isDir(const std::string &name) const;

  /// Get info about a single file
  virtual FileInfo stat(const std::string &name) const;

  /// List all entries in a given directory. A blank dir should be
  /// interpreted as a the root/current directory of the archive. If
  /// dirs is true, list directories instead of files.
  virtual FileInfoList list(const std::string& dir = "",
                            bool recurse=true,
                            bool dirs=false) const;

  /// Find files after a given pattern. Wildcards (*) are
  /// supported. Only valid if 'hasFind' is true. Don't implement your
  /// own pattern matching here if the backend doesn't support it
  /// natively; use a filter instead (not implemented yet.)
  virtual FileInfoList find(const std::string& pattern,
                            bool recursive=true,
                            bool dirs=false) const;
};

}} // namespaces
#endif
