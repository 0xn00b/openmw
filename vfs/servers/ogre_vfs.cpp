#include "ogre_vfs.h"
#include "../../stream/servers/ogre_datastream.h"

using namespace Mangle::VFS;

OgreVFS::OgreVFS(const std::string &_group)
  : group(_group)
{
  hasList = true;
  hasFind = true;
  isCaseSensitive = true;

  // Get the group manager once
  gm = Ogre::ResourceGroupManager::getSingletonPtr();

  // Use the default group if none was specified
  if(group.empty())
    group = gm->getWorldResourceGroupName();
}

Mangle::Stream::Stream *OgreVFS::open(const std::string &name)
{
  Ogre::DataStreamPtr data = gm->openResource(name, group);
  return new Stream::OgreStream(data);
}

static void fill(FileInfoList &out, Ogre::FileInfoList &in, bool dirs)
{
  int size = in.size();
  out.resize(size);

  for(int i=0; i<size; i++)
    {
      out[i].name = in[i].filename;
      out[i].basename = in[i].basename;
      out[i].size = in[i].uncompressedSize;
      out[i].isDir = dirs;
      out[i].time = 0;
    }
}

FileInfoList OgreVFS::list(const std::string& dir,
                            bool recurse,
                            bool dirs) const
{
  Ogre::FileInfoListPtr olist = gm->listResourceFileInfo(group, dirs);
  FileInfoList res;
  fill(res, *olist, dirs);
  return res;
}

FileInfoList OgreVFS::find(const std::string& pattern,
                            bool recursive,
                            bool dirs) const
{
  Ogre::FileInfoListPtr olist = gm->findResourceFileInfo(group, pattern, dirs);
  FileInfoList res;
  fill(res, *olist, dirs);
  return res;
}
