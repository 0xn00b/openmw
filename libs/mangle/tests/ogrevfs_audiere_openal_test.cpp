/*
  This example combines:

  - the OGRE VFS system (to read from zip)
  - Audiere (for decoding sound data)
  - OpenAL (for sound playback)

 */

#include "sound/filters/openal_audiere.hpp"
#include "vfs/servers/ogre_vfs.hpp"
#include <Ogre.h>
#include <iostream>

using namespace Ogre;
using namespace Mangle;
using namespace std;

int main()
{
  // Disable Ogre logging
  new LogManager;
  Log *log = LogManager::getSingleton().createLog("");
  log->setDebugOutputEnabled(false);

  // Set up Root
  Root *root = new Root("","","");

  // Add zip file with a sound in it
  root->addResourceLocation("sound.zip", "Zip", "General");

  // Ogre file system
  VFS::OgreVFS vfs;

  // The main sound system
  Sound::OpenAL_Audiere_Factory mg;
  Sound::SoundPtr snd = mg.load(vfs.open("owl.ogg"));

  cout << "Playing 'owl.ogg' from 'sound.zip'\n";
  snd->play();

  while(snd->isPlaying())
    {
      usleep(10000);
      if(mg.needsUpdate) mg.update();
    }

  return 0;
}
