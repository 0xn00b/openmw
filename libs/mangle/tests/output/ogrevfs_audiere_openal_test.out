Playing 'owl.ogg' from 'sound.zip'
